src/main/java/org/apache/commons/crypto/cipher/Openssl.java

```diff
@@ -43,9 +49,20 @@
     public static final int ENCRYPT_MODE = 1;
     public static final int DECRYPT_MODE = 0;
 
+    static final int DEFAULT_TAG_LEN = 16;
+
+    private int mode = Openssl.DECRYPT_MODE;
+
+    // buffer for AAD data; if consumed, set as null
+    private byte[] aadBuffer;
+    private int tagBitLen = -1;
+
+    // buffer for storing input in decryption, not used for encryption
+    private ByteArrayOutputStream ibuffer = null;
```

Comment:
According to GCM Spec, after the AAD is processed, GCM cipher got a initial intermediate tag value (denoted as X in the Spec).  Then, X is XORed with the ciphertext blocks one by one.
From my understanding, Java doc means that, if the initial intermediate tag value is NOT ready( AAD is not processed) before encryption(update/doFinal), to get the intermediate tag value, GCM cipher has to internally buffer ciphertext for later calculation of  intermediate tag value.  If AAD data is ready, GCM cipher only has to buffer the intermediate tag value.

In a word, gcm implementation has to avoid buffering ciphertext not the AAD.

Thanks for your reminder, I find a bug. we can not use byte[] as buffer, because updateAAD()  can be called zero or more times. the aad data should be concatenated. it seems that ByteArrayOutputStream is suitable for buffering aad. I will update soon.
```java
//updateAAD()  can be called zero or more times
cipher.updateAAD(aad1);
cipher.updateAAD(aad2);
```